File: python/medium/Validate_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def next_node(node):
            val = node.val
            if node.right is None and node.left is None:
                return [node]
            else:
                if node.left is not None:
                    left_list = next_node(node.left)
                    if not left_list:
                        return False
                    for left_node in left_list:
                        if val <= left_node.val:
                            return False
                else:
                    left_list = []
                if node.right is not None:
                    right_list = next_node(node.right)
                    if not right_list:
                        return False
                    for right_node in right_list:
                        if val >= right_node.val:
                            return False
                else:
                    right_list = []
                return left_list + [node] + right_list
        if not next_node(root):
            return False
        else:
            return True

File: cpp/medium/Validate_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        vector<int> left;
        vector<int> right;
        return nextNode(root, left, right);
    }
    bool nextNode(TreeNode* node, vector<int>& left, vector<int>& right) {
        for (auto i:left) {
            if (node->val >= i) return false;
        }
        for (auto i:right) {
            if (node->val <= i) return false;
        }
        if (node->left != nullptr) {
            left.push_back(node->val);
            if (!nextNode(node->left, left, right)) return false;
            left.pop_back();
        }
        if (node->right != nullptr) {
            right.push_back(node->val);
            if (!nextNode(node->right, left, right)) return false;
            right.pop_back();
        }
        return true;
    }
};
